crates/test_utils/src/shared_values/mod.rs

```diff
@@ -0,0 +1,1011 @@
+    pub(crate) const TEST_AGENT_READINESS_TIMEOUT_SECS_DEFAULT: u64 = 360;
+
+    #[derive(Debug, Default, strum_macros::EnumString)]
+    #[strum(serialize_all = "lowercase")]
+    pub(crate) enum RemoteV1Role {
+        #[default]
+        Both,
+
+        Server,
+        Client,
+    }
+
+    /// Remote implementation using Websockets for data passing.
+    #[derive(Clone)]
+    pub(crate) struct RemoteV1Server {
+        local_addr: url2::Url2,
+
+        server_handle: Arc<JoinHandle<Fallible<()>>>,
+    }
+
+    /// Payload for requests
+    #[derive(Serialize, Deserialize, SerializedBytes, Debug, Clone)]
+    pub(crate) enum RequestMessage {
+        Test(String),
+        Put {
+            key: String,
+            value: String,
+        },
+        Get {
+            pattern: String,
+            min_data: usize,
+            maybe_timeout: Option<Duration>,
+        },
+        NumWaiters,
+    }
+
+    /// Payload for responses
+    #[derive(Serialize, Deserialize, SerializedBytes, Debug, Clone)]
+    pub(crate) enum ResponseMessage {
+        StringErr(String),
+        Test(String),
+        Put(Result<Option<String>, String>),
+        Get(Result<Data<String>, String>),
+        NumWaiters(usize),
+    }
+
+    impl RemoteV1Server {
+        /// Creates a new server and starts it immediately.
+        pub(crate) async fn new(bind_socket: Option<String>) -> Fallible<Self> {
+            let localv1 = LocalV1::default();
+
+            let original_url = url2::Url2::try_parse(bind_socket.unwrap_or_else(|| {
+                std::env::var(SHARED_VALUES_REMOTEV1_URL_ENV)
+                    .map_err(|e| {
+                        tracing::debug!(
+                            "could not read env var {SHARED_VALUES_REMOTEV1_URL_ENV}: {e}"
+                        );
+                        e
+                    })
+                    .ok()
+                    .unwrap_or(SHARED_VALUES_REMOTEV1_URL_DEFAULT.to_string())
+            }))?;
+
+            tracing::debug!("binding server to {original_url}");
+
+            let mut server = WebsocketListener::bind(
+                original_url.clone(),
+                std::sync::Arc::new(WebsocketConfig::default()),
+            )
+            .await?;
+
+            let local_addr = server.local_addr().clone();
+
+            tracing::info!("server bound to {local_addr}");
+
+            let server_handle = tokio::task::spawn(async move {
+                // Handle new connections
+                Self::remotev1server_inner(localv1, &mut server).await?;
+
+                Ok(())
+            });
+
+            Ok(Self {
+                local_addr,
+                server_handle: Arc::new(server_handle),
+            })
+        }
+
+        async fn remotev1server_inner(
+            localv1: LocalV1,
+            server: &mut WebsocketListener,
+        ) -> Fallible<()> {
+            while let Some(Ok((/* never sends on its own */ _tx, mut recv))) = server.next().await {
```

Comment:
is `_tx` safe to drop though? Because this loop spawns tokio tasks which means this will go out of scope quite quickly